Add new Roborock Integration

Lash-L opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The current Roborock integration which is just an alias for Xiaomi Miio will no longer be an alias, but instead will be this integration. I'm not sure what the best way to handle this is. In my opinion, a dedicated Roborock integration should take priority over an alias integration. But I'm not sure how that will affect current users

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This is a conversion of the popular hacs integration: https://github.com/humbertogontijo/homeassistant-roborock. It aims to allow users to be able to use the Roborock app and the home assistant integration, rather than having to use the Xiaomi app. The roborock app gives a significantly better use experience than the Xiaomi app for Roborock users. The aim is also to provide better support for Roborock specific devices.

Because this is a HACS conversion, the original project was very large. I tried to remove things that would not be needed for an initial PR. Some files are still rather larger as it was difficult to determine what was worth removing to keep the PR small and what was functionality that it didn't make sense to remove( A large portion of this PR is from test on vacuum.py. I could potentially remove them to keep the PR smaller, but I wasn't sure if it made sense)

Here are the features that exist that were removed for this PR and will be added in future PRS once this is merged in:
- Email/Password login
- Map support
- Binary Sensors
- Sensors
- Option Flow
- Manual Vacuum control
- go to x y location
- mopping
- services

This integration uses the following pypi package: https://github.com/humbertogontijo/python-roborock

As a note, the mock data is around 1/3 of this pr, so it is not quite as big as it originally seems

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/26520

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Lash-L committed:
init commit of roborock

Lash-L committed:
removed some non-vacuum related code

Lash-L committed:
removed some non-needed constants

Lash-L committed:
removed translations

Lash-L committed:
removed options flow

Lash-L committed:
removed manual control

Lash-L committed:
remove password login

Lash-L committed:
removed go-to

Lash-L committed:
removed unneeded function and improved device_stat

Lash-L committed:
removed utils as it is unused

Lash-L committed:
typing changes in vacuum.py

Lash-L committed:
fixed test patch paths

Lash-L committed:
removed unneeded records

Lash-L committed:
removing unneeded code in tests

Lash-L committed:
remove password from strings

Lash-L committed:
removed maps in code

Lash-L committed:
changed const, reworked functions

Lash-L committed:
remove menu

Lash-L committed:
fixed tests

Lash-L committed:
100% code coverage config_flow

Lash-L committed:
init roborock commit

Lash-L committed:
Merge branch 'home-assistant:dev' into roborock_comparision

Lash-L committed:
small changes

Lash-L committed:
removed unneeded patch

Lash-L committed:
bump to 0.1.7

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

frenck reviewed (dismissed):
Please make PRs small and limit them to the smallest usable set. For example, remove the additional services, those can be added later.

See also: 

https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Lash-L committed:
removed services

Lash-L committed:
Merge branch 'roborock_hacs_conversion' of https://github.com/Lash-L/core into roborock_hacs_conversion

Lash-L committed:
removed extra functions and mop

Lash-L commented:
> Please make PRs small and limit them to the smallest usable set. For example, remove the additional services, those can be added later.
> 
> See also:
> 
> https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Understood, I initially was unsure if I should have kept services in, but I removed them + mopping + tests that used either of those.

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

Lash-L committed:
add () to configEntryNotReady

Lash-L committed:
moved coordinator into seperate file

Lash-L committed:
update roborock testing

Lash-L committed:
removed stale options code

Lash-L committed:
normalize username for unique id

Lash-L committed:
removed unneeded variables

Lash-L committed:
fixed linter problems

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

Lash-L committed:
removed stale comment

allenporter commented on homeassistant/components/roborock/strings.json:
```diff
@@ -0,0 +1,34 @@
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
+      "already_in_progress": "Account is already being configured"
```
Are these abort codes actually supported by this integration?

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,147 @@
+        self._client: RoborockClient | None = None
+        self._auth_method: str | None = None
+
+    async def async_step_user(
```
These forms show one step and switch between another step, which is not allowed. While rendering step user it sends the user to step email, and instead it needs to stay on step user. These steps should be combined.  To advance to the next step, call the next step and await on it.

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+    @property
+    def supported_features(self) -> VacuumEntityFeature:
+        """Flag vacuum cleaner features that are supported."""
+        features = (
+            VacuumEntityFeature.PAUSE
+            | VacuumEntityFeature.STOP
+            | VacuumEntityFeature.RETURN_HOME
+            | VacuumEntityFeature.FAN_SPEED
+            | VacuumEntityFeature.BATTERY
+            | VacuumEntityFeature.STATUS
+            | VacuumEntityFeature.SEND_COMMAND
+            | VacuumEntityFeature.LOCATE
+            | VacuumEntityFeature.CLEAN_SPOT
+            | VacuumEntityFeature.STATE
+            | VacuumEntityFeature.START
+        )
+        return features
```
```suggestion
        return (
            VacuumEntityFeature.PAUSE
            | VacuumEntityFeature.STOP
            | VacuumEntityFeature.RETURN_HOME
            | VacuumEntityFeature.FAN_SPEED
            | VacuumEntityFeature.BATTERY
            | VacuumEntityFeature.STATUS
            | VacuumEntityFeature.SEND_COMMAND
            | VacuumEntityFeature.LOCATE
            | VacuumEntityFeature.CLEAN_SPOT
            | VacuumEntityFeature.STATE
            | VacuumEntityFeature.START
        )
```

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+    @property
+    def icon(self) -> str:
+        """Return the icon of the vacuum cleaner."""
+        return "mdi:robot-vacuum"
```
This could be an `_attr_icon`

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+        """Initialize a vacuum."""
+        StateVacuumEntity.__init__(self)
+        RoborockCoordinatedEntity.__init__(self, device, coordinator, unique_id)
+        self.manual_seqnum = 0
```
This appears to be unused and should be removed

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+
+    def __init__(
+        self,
+        unique_id: str,
```
Any particular reason to change the initialization order between this class and the parent class? I'd opt to keep it the same (either change this or the other)

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+        StateVacuumEntity.__init__(self)
+        RoborockCoordinatedEntity.__init__(self, device, coordinator, unique_id)
+        self.manual_seqnum = 0
+        self._device = device
```
The parent class takes every field of `_device` and copies it as a member variable. How about picking one apparoch other the other.  This appears to be unused.

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,233 @@
+        RoborockCoordinatedEntity.__init__(self, device, coordinator, unique_id)
+        self.manual_seqnum = 0
+        self._device = device
+        self._coordinator = coordinator
```
The base coordinator already has `self.coordinator` so i don't think this is needed

allenporter commented on homeassistant/components/roborock/manifest.json:
```diff
@@ -1,6 +1,9 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/roborock",
+  "iot_class": "cloud_polling",
+  "requirements": ["python-roborock==0.1.7"]
```
Looking at this on https://pypi.org/project/python-roborock/ it seems like this project needs a little more work (e.g. has TODOs, no links to github, etc) so can you do some work to get this into a healthier state? I ended up doing some digging and found https://github.com/humbertogontijo/python-roborock which appears to be the source and i see you're doing cleanup work. (I was surprised this was running such an an older vesion with upates in the last 24 hours)

Lash-L replied:
Yeah good point - I'll take some time to work on the pypi page tomorrow.

As far as the version being delayed, offline support is currently being added to the package, so there's a lot of rapid development, I was going to wait until we had it fully tested and functionaly before it got bumped

allenporter reviewed (changes requested):
Hi, I tried to give this a first pass, but did not review the whole thing. In general, I would suggest taking a pass to simplify and remove dead or unnecessary code.

I am not familiar with the best practices these days around brands and multiple integrations vs single integrations and how existing users need to be handled -- but generally I agree that going through the xiaomi app is not good and using a native roborock integration would make a lot more sense to me assuming there aren't any type of migration down sides etc. I think i'm not the right person to make this call.

Lash-L committed:
additional pr changes

Lash-L commented:
Thanks for the feedback Allen. I handled some of the more simple tasks, and I'll spend some time tomorrow taking care of the more difficult tasks, i.e. pypi and config flow rework and giving the code another pass to try to find anything unused.

Lash-L committed:
simplify config_flow

Lash-L committed:
fix config flow test

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

allenporter reviewed (changes requested):
Thanks for the improvements. I took another pass and requested some changes below.

This is some optional feedback: I was having a look at the python library and noticed that it has a lot of classes that subclass `dict` and then provides fields that are accessors. You may consider using pydantic or another library like that to aid with response validation as it also really helps with typing, etc.  I think it also help simplify cases where this integration is passing around empty typed dictionaries which may or may not contain valid values.  This is totally up to you, but i thought i would mention this as its a popular choice.

Lash-L committed:
Apply suggestions from code review

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Lash-L commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -0,0 +1,70 @@
+        )
+        device_map[device.duid] = RoborockDeviceInfo(device, product)
+
+    client = RoborockMqttClient(user_data, device_map)
```
Yeah - pushing is definitely better - I'll make that change

allenporter replied:
To clarify: I was asking a different question: is this actually polling or subscribing? The client library is called mqtt but it appears to be polling.... If it really is subscribing, then its not clear you need to poll for latest device state so often.  I wasn't saying to change the manifest, but hoping for clarification on what is actually happening.

I also reviewed the library and it does not seem to implement async correctly. For example, in the shutdown method it just calls the sync methods form inside async. I have some serious concerns about this client library being well behaving asyncio.  Does my concern make sense or am I missing something?

Lash-L replied:
I asked the original creator to clarify - as that's all his code and he would be better answering any questions

Lash-L committed:
First pass at resolving PR comments

Lash-L committed:
Merge branch 'roborock_hacs_conversion' of https://github.com/Lash-L/core into roborock_hacs_conversion

Lash-L commented on tests/components/roborock/test_config_flow.py:
```diff
@@ -0,0 +1,148 @@
+
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["step_id"] == "code"
+    # Raise exception for invalid code
```
Okay added continuing after error. I basically have one large helper method that checks everything now for various errors and patches.

Lash-L committed:
reworked config flow

Lash-L commented:
Thanks for the second pass Allen! I've gone through and made the changes, I'll make a note of pydantic and potentially add it down the line

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,180 @@
+        """Initialize a vacuum."""
+        StateVacuumEntity.__init__(self)
+        RoborockCoordinatedEntity.__init__(self, unique_id, device, coordinator)
+        self._attr_icon = "mdi:robot-vacuum"
+        self._attr_supported_features = (
```
These constants can be attributes at the class level

allenporter commented on tests/components/roborock/test_config_flow.py:
```diff
@@ -0,0 +1,108 @@
+from .mock_data import MOCK_CONFIG, USER_EMAIL
+
+
+async def config_flow_helper(
```
Test helpers like this make tests harder to read and are discouraged. Instead, use either test fixtures, parameterized tests, or just duplicate code between tests.

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        self._errors.clear()
```
Now that everything is inlined errors can be a local variable so no need to clear each step.

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+            await self.async_set_unique_id(username.lower())
+            self._abort_if_unique_id_configured()
+            self._username = username
+            assert self._username is not None
```
This can't happen given username is a required string. We generally don't want assertions in here.

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+
+        if user_input is not None:
+            code = user_input[CONF_ENTRY_CODE]
+            user_data = await self._code_login(code)
```
In the last round, and again here, I don't think the helper methods are super helpful here given that it requires using side effects on the class like populating `self._errors`. Inline this and remove `_errors` as a member variable, and just keep it as a local variable here and in the prior step.

As an example, there are a lot of assertions here and many of them could be removed given the variables would already referenced as local variables in this function.

Lash-L replied:
Are you recommending combining step_user and step_code?

I'm not sure how I could avoid asserting self._username here and self._client, as I can't pass them through the steps.

I don't think I can combine user and code, as I need the user's email before a 2fa code can be shown so I need two different forms. Open to any suggestions though. I think I at least (slightly?) improved it

allenporter replied:
Sorry let me clarify: You still need two steps. I was saying to inline `_show_code_form` and `_code_login` `_create_entry`.

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        self._username: str | None = None
+        self._errors: dict[str, str] = {}
+        self._client: RoborockClient | None = None
+        self._auth_method: str | None = None
```
Appears to be unused

allenporter reviewed (changes requested):
My main concern now is about the client library on two fronts:
- is it actually implementing async correctly? It appears to just do blocking calls in async methods
- is it polling or subscribing? it seems to start a subscription then ignore it to just poll

humbertogontijo commented:
@allenporter what do you mean by actually implementing async correctly? The library have everything asynchronous just to not block the main thread waiting for the mqtt response.
The api uses mqtt, but the device do not publish status unless a mqtt message is sent. So its polling unfortunately

Lash-L committed:
moving vacuum attr

Lash-L committed:
attempt to clean up conflig flow more

allenporter commented:
> @allenporter what do you mean by actually implementing async correctly? The library have everything asynchronous just to not block the main thread waiting for the mqtt response. The api uses mqtt, but the device do not publish status unless a mqtt message is sent. So its polling unfortunately

OK i may be misunderstanding. These `async_connect` and `async_disonnect` methods appear to just call the sync methods [example](https://github.com/humbertogontijo/python-roborock/blob/17e72c34c6ac133025450eab68f4be7025ab138b/roborock/cloud_api.py#L146) on the mqtt client.

humbertogontijo commented:
Yes, Paho mqtt uses callbacks for both connection and disconnection, so async_connect and async_disconnect methods will call the sync method, and await for the callback to be called

humbertogontijo commented:
> > @allenporter what do you mean by actually implementing async correctly? The library have everything asynchronous just to not block the main thread waiting for the mqtt response. The api uses mqtt, but the device do not publish status unless a mqtt message is sent. So its polling unfortunately
> 
> OK i may be misunderstanding. These `async_connect` and `async_disonnect` methods appear to just call the sync methods [example](https://github.com/humbertogontijo/python-roborock/blob/17e72c34c6ac133025450eab68f4be7025ab138b/roborock/cloud_api.py#L146) on the mqtt client.

Yes, Paho mqtt uses callbacks for both connection and disconnection, so async_connect and async_disconnect methods will call the sync method, and await for the callback to be called

allenporter commented:
> > > @allenporter what do you mean by actually implementing async correctly? The library have everything asynchronous just to not block the main thread waiting for the mqtt response. The api uses mqtt, but the device do not publish status unless a mqtt message is sent. So its polling unfortunately
> > 
> > 
> > OK i may be misunderstanding. These `async_connect` and `async_disonnect` methods appear to just call the sync methods [example](https://github.com/humbertogontijo/python-roborock/blob/17e72c34c6ac133025450eab68f4be7025ab138b/roborock/cloud_api.py#L146) on the mqtt client.
> 
> Yes, Paho mqtt uses callbacks for both connection and disconnection, so async_connect and async_disconnect methods will call the sync method, and await for the callback to be called

OK thanks. So the sync methods don't actually do any I/O and you have to call `loop()` to do the IO? Interesting API! :)

Thanks for confirming.

allenporter reviewed (changes requested):
@Lash-L so if its polling then seems like you don't want to change the type to push? Also there were some other comments resolved but i don't see the changes yet.

Lash-L commented:
@allenporter Yeah,  I have some local changes in which I am changing it back to polling.

Which comments in particular? I thought I resolved them all and I just did a quick look through (I could have easily missed things) and it looked like I fixed them. I made all of the changes to config flow and removed all the unneeded assertions, reworked the config flow tests, and all of the other comments you made. If there is anything you notice that I didn't resolve, just unresolve the comment and I'll take a look - I'm sorry about that.

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+            self._abort_if_unique_id_configured()
+            self._username = username
+            try:
+                _LOGGER.debug("Requesting code for Roborock account")
+                client = RoborockClient(username)
+                self._client = client
```
These should be moved outside of the `try` since they do not raise the exceptions

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+            self._username = username
+            try:
+                _LOGGER.debug("Requesting code for Roborock account")
+                client = RoborockClient(username)
+                self._client = client
+                await client.request_code()
```
Worth saving 1 line:

```suggestion
                self._client = RoborockClient(username)
                await self._client.request_code()
```

allenporter commented on homeassistant/components/roborock/config_flow.py:
```diff
@@ -0,0 +1,100 @@
+        if user_input is not None:
+            code = user_input[CONF_ENTRY_CODE]
+            try:
+                _LOGGER.debug("Logging into Roborock account using email provided code")
```
Can be moved outside of the scope of the try block

allenporter commented on homeassistant/components/roborock/vacuum.py:
```diff
@@ -0,0 +1,181 @@
+        return self._device_status.error
+
+    @property
+    def capability_attributes(self) -> dict[str, list[str]]:
```
This seems to just be duplicating code in the parent class and is not needed. (Plus `FAN_SPEED` is always present so its more complicated than needs to be)

allenporter commented on tests/components/roborock/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+        )
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
+    if request_code_side_effect:
```
You've made good improvements here based on the fixtures, etc However one thing to further clarify is that we don't want conditionals in tests like this.  

So to be more specific:
- have one test that fails based on request code failures
- have one test that fails based on login code failures

The motivation behind this is we'd rather have some code duplication, but easier to follow tests.

I also think you want to add back a test that completes the whole flow without errors from the start.

allenporter reviewed (changes requested):
Thanks for addressing all the previous feedback. (You got everything I previously asked for, thanks!)

I think this is the last round and otherwise this looks good to go with just a few small items.  I'm very excited to use this myself as i was turned off by the existing integration since its too complicated to setup and this appears much easier. 👍🏼

Lash-L commented:
> Thanks for addressing all the previous feedback. (You got everything I previously asked for, thanks!)
> 
> I think this is the last round and otherwise this looks good to go with just a few small items. I'm very excited to use this myself as i was turned off by the existing integration since its too complicated to setup and this appears much easier. 👍🏼

Awesome! I'll take a look at these and bundle them with the changes I'm working on.

I'm not sure if you saw my edit above - but I was planning to update this to be a local polling integration from the start. So I bumped up the version.

It still uses an online connection to get your account information and then all of your devices and the local key and the ip of the devices, but then it communicates entirely through local connections. I have a feeling you will have to give a little bit more feedback once I put that out.

If you think it would be better - I could hold off on making the change to make it offline and instead just get this initial version out with it being online only.

allenporter commented:
> > Thanks for addressing all the previous feedback. (You got everything I previously asked for, thanks!)
> > I think this is the last round and otherwise this looks good to go with just a few small items. I'm very excited to use this myself as i was turned off by the existing integration since its too complicated to setup and this appears much easier. 👍🏼
> 
> Awesome! I'll take a look at these and bundle them with the changes I'm working on.
> 
> I'm not sure if you saw my edit above - but I was planning to update this to be a local polling integration from the start. So I bumped up the version.

Oh I missed that.

> 
> It still uses an online connection to get your account information and then all of your devices and the local key and the ip of the devices, but then it communicates entirely through local connections. I have a feeling you will have to give a little bit more feedback once I put that out.
> 
> If you think it would be better - I could hold off on making the change to make it local and instead just get this initial version out with it being online only.

Up to you, i don't mind either way. I think the important thing to worry about here would just be the config flow details and if those are saying the same, then i'm fine with either approach.

Lash-L commented:
> Up to you, i don't mind either way. I think the important thing to worry about here would just be the config flow details and if those are saying the same, then i'm fine with either approach.

Okay I'm going to add local then for this release.

Here's my logic - if you have a chance let me know if I'm doing something that breaks standards.

* config flow would store the Network information that is needed for each device that you have
* In __init__, we attempt to use the cloud connected mqttclient to update network information/ get any new devices
* If a new devices is found - are we allowed to add the information to the ConfigEntry? Or are we not supposed to change the config entry in init? 
* If the mqtt client can't get the networking information, we just use the stored local information.

My thought with this approach is that someone could block their roborock from reaching outside their home network and everything would still work correctly. Then if they got another Roborock as long as it isn't blocked for the first time you open up Homeassistant, it would be added, and then you could block it.

allenporter commented:
> edit edit: You can ignore this for now and just wait for my implementation - I think I figured it out.
> 
> > Up to you, i don't mind either way. I think the important thing to worry about here would just be the config flow details and if those are saying the same, then i'm fine with either approach.
> 
> Okay I'm going to add local then for this release.
> 
> Here's my logic - if you have a chance let me know if I'm doing something that breaks standards.
> 
> * config flow would store the Network information that is needed for each device that you have
> * In **init**, we attempt to use the cloud connected mqttclient to update network information/ get any new devices
> * If a new devices is found - are we allowed to add the information to the ConfigEntry? Or are we not supposed to change the config entry in init?
> * If the mqtt client can't get the networking information, we just use the stored local information.
> 
> My thought with this approach is that someone could block their roborock from reaching outside their home network and everything would still work correctly. Then if they got another Roborock as long as it isn't blocked for the first time you open up Homeassistant, it would be added, and then you could block it.
> 
> If I just handled everything within the init, then I wouldn't be able to store(unless I am able to add to config entry?) , and then the roborock would have to be reachable on the internet everytime you started up HomeAssistant
> 
> Edit: Maybe for an initial approach it just makes sense to have it communicate locally but not support it being blocked completely from the outside?
> 
> What are your thoughts?
> 
> Edit edit: I found hass.config_entries.async_update_entry so I am using that in init and not modifying config flow at all

Given this is the initial integration, i would suggest to keep this as simple as possible, either:
- get the list of devices on startup and use that
- get the list of devices from the cloud in the config flow and use that

The other cases seem like they are optimizations or improvements that can follow in future smaller PRs. (If you go with the first approach, its  improving reliability by using falling back to stored configuration when the cloud is not available.  If you go with the second approach, its improving device discovery, automatically discovering and setting up new devices.). This way we can focus on getting something initially checked in. This will also allow having reviews in parallel for future platforms with discovery improvements.

Lash-L committed:
update package and use offline functionality

Lash-L commented:
> The other cases seem like they are optimizations or improvements that can follow in future smaller PRs. (If you go with the first approach, its improving reliability by using falling back to stored configuration when the cloud is not available. If you go with the second approach, its improving device discovery, automatically discovering and setting up new devices.). This way we can focus on getting something initially checked in. This will also allow having reviews in parallel for future platforms with discovery improvements.


Understood! For now I'll just get it in init then. and I will add the extra functionality in a future PR. 

I've pushed the changes, I assume I'll need some changes, so just let me know whatever you think you should. There is one bug with the fan_speed we are working on, but other than that I have tested it a bit with some pretty good success. 

If we can't figure out the fan_speed bug, I'm going to switch this back to being online polling as the bug does not exist there.

Lash-L committed:
Fixed errors and fan bug

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

Lash-L commented:
Fixed issues with the tests and the fan speed bug.

allenporter commented on tests/components/roborock/test_config_flow.py:
```diff
@@ -8,28 +8,118 @@
+    hass: HomeAssistant,
+    bypass_api_fixture,
+) -> None:
+    """Handle applying errors to request code or code login and recovering from the errors."""
```
Update doc string

allenporter commented on tests/components/roborock/test_config_flow.py:
```diff
@@ -8,28 +8,118 @@
+) -> None:
+    """Handle applying errors to request code or code login and recovering from the errors."""
+    with patch(
+        "homeassistant.components.roborock.async_setup_entry", return_value=True
```
My impression is that this is method is not actually called here, but later once the config entry is set up and created when RESULT_TYPE_CREATE_ENTRY is returned. (here and below). I think the common practice i see is to name this match (e.g. mock_setup) and assert that it was actually called at the end of the test.

allenporter commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -24,27 +31,37 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
-    device_map: dict[str, RoborockDeviceInfo] = {}
-    devices = home_data.devices + home_data.received_devices
+    devices_info: dict[str, RoborockDeviceInfo] = {}
```
There is a typing issue here.  This is `dict[str, RoborockDeviceInfo]` but the coordinator expects `dict[str, RoborockHassDeviceInfo]` i think.

Plus, creating a dict then either deleting entries or replacing entries with a different type is harder to follow. I'd say just iterate over the devices and populate the final dict of  `RoborockHassDeviceInfo`.

Make products into a dict as well to simplify and shorten the loop

Lash-L replied:
I think I need three separate dicts, especially after switching to composition.

Mqtt Client takes a dict of device info

Local Client takes a dict of local device info

Then the coordinator takes a dict of hass device info

Is there a better way that I am missing?

allenporter replied:
I don't see anything super obvious you're missing, but here are some ideas:
- inline some of the dict creation in the client library constructors
- push the client libraries creation all into the update coordinator (e.g. the coordinator just takes a list of devices as input, and hides all the client library details from init. doesn't necessarily solve the issue, but you may find that this gets easier to manage in one place).   for example, do you still want to create the entities anyway and just mark them `unavailable` rather than not creating them at all?
- simplify the client library constructors e.g. the client library itself takes a dict, but it could just also take a list of devices
- generally simplify the client library to abstract all of this away to provide the APIs you actually want to use

Lash-L replied:
Moved it into constructor and inlined some of the dict creation, could potentially inline inside the coordinator when I create the local_devices_info and hass_devices_info, but because they need networking, I wasn't sure if it was efficient, as they would both end up reading from the networking dict as well.

allenporter commented on homeassistant/components/roborock/models.py:
```diff
@@ -0,0 +1,12 @@
+
+
+@dataclass
+class RoborockHassDeviceInfo(RoborockLocalDeviceInfo):
```
Optional: Preferring composition to inheritance would probably improve readability  since this is subclassing a dataclass in another package

allenporter commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -23,34 +24,38 @@ class RoborockDataUpdateCoordinator(
-            return self._devices_prop
-        raise UpdateFailed("No device props found")
+        return {
+            device_id: device_info.props
```
`device_info.props` can be `None` which seems to conflict with the return type which does not allow `None`.  Is there any value in having a none `RoborockDeviceProp` or should it just always be initialized with empty since the values themselves can still be none?

(Aside: though confusingly the typing on the fields in `RoborockDeviceProp` is typed as if they can't be None, but they can be.  Generally, typing in the library seems wrong in many places)

Lash-L replied:
Yeah we need to setup mypy in the package to catch the typing issues. I've fixed this specific issue for now.

allenporter commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -23,34 +24,38 @@ class RoborockDataUpdateCoordinator(
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: RoborockMqttClient,
```
I believe this is not the correct type

Lash-L committed:
rework model and some other small changes

Lash-L committed:
bump version

allenporter commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -37,31 +38,29 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-            ),
-            {},
+        hass_devices_info[device.duid] = RoborockHassDeviceInfo(
+            device, networking, product_info[device.product_id]
```
Is `device.product_id` guaranteed to be present in `product_info`?

Lash-L replied:
Yes - it should be.

Lash-L committed:
used default factory

allenporter commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -37,31 +38,29 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     devices: list[HomeDataDevice] = home_data.devices + home_data.received_devices
     for device in devices:
-        devices_info[device.duid] = RoborockDeviceInfo(device)
+        mqtt_devices_info[device.duid] = RoborockDeviceInfo(device)
```
I think you can also shorten to something like this:
```
mqtt_devices_info = {
    device.duid: RoborockDeviceInfo(device)
    for device in devices
}
```

or even inline further as an argument to `RoborockMqttClient`

Lash-L committed:
moved some client creation into coord

Lash-L committed:
fixed patch

allenporter commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -0,0 +1,65 @@
+        user_data, {device.duid: RoborockDeviceInfo(device) for device in devices}
+    )
+    network_info = {
+        device.duid: await mqtt_client.get_networking(device.duid) for device in devices
```
Sorry for the late comment -- both here and in the update coordinator, when doing multiple serial awaits the preference is instead to use `asyncio.gather`

Lash-L replied:
No need to apologize. I've made the change

allenporter commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -0,0 +1,84 @@
+        local_devices_info: dict[str, RoborockLocalDeviceInfo] = {}
+        hass_devices_info: dict[str, RoborockHassDeviceInfo] = {}
+        for device in devices:
+            networking = devices_networking.get(device.duid)
+            if networking is None:
```
```suggestion
            if not (networking := devices_networking.get(device.duid):
```

Lash-L committed:
Update homeassistant/components/roborock/coordinator.py

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Lash-L committed:
moved async functions into gather

allenporter commented on homeassistant/components/roborock/__init__.py:
```diff
@@ -35,7 +36,13 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     network_info = {
-        device.duid: await mqtt_client.get_networking(device.duid) for device in devices
+        device.duid: result
+        for device, result in zip(
```
Now that its bigger and more than one line the style prefers not using a comprehension.

Lash-L replied:
That change better?

allenporter commented on homeassistant/components/roborock/coordinator.py:
```diff
@@ -73,8 +74,12 @@ async def _update_device_prop(self, device_info: RoborockHassDeviceInfo) -> None
-            for device_info in self.devices_info.values():
-                await self._update_device_prop(device_info)
+            asyncio.gather(
+                *[
```
I think it is preferred to use a tuple instead of a list when a list is not needed.

Lash-L committed:
reworked gathers

Lash-L committed:
removed random line

allenporter reviewed (approved):
Looks good @Lash-L 👍🏼

Lash-L commented:
did you cancel the codecov update Allen or did it just get cancelled based on time?

I was gonna update branch to get it to pass, but I guess that doesn't really matter?

Lash-L committed:
error catch if networking doesn't exist or timeout

Lash-L commented:
I realized there was a potential problem where if you timed out getting the networking information or the networking wasn't available, you could be left with 0 devices and technically the integration would still set up. I believe a ConfigEntryNotReady is more appropriate.

Lash-L committed:
bump to 0.6.5

Lash-L commented:
Hey allen, sorry to make another push, but I went through and fixed typing in the base package using mypy - so this iversion bump should theoretically make us slightly less likely to get errors so I wanted to include it.

Let me know if there is anything else you need me to do. I won't push anymore unless this pipeline fails.

Lash-L committed:
fixed mocked data reference url

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

allenporter reviewed (approved):
I re-reviewed the updates form the python library and those look good. Appreciate the effort put into cleanup and all the typing improvements, etc. This integration could likely be straight forward to move to strict typing, which I would encourage. There may be additional feedback from other reviewers once merged, which can be addressed in future PRs, and i know you plan to keep porting in new functionality, which i'd be happy to try to find and review. 

I have one small nit, then happy to merge.

Lash-L committed:
change checking if we have no network information

Co-authored-by: Allen Porter <allen.porter@gmail.com>

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

Lash-L commented:
Sounds good - I'll add strict typing down the line. I ran it now just to see, and there are a few things I have to change in the package first, so I'll add it to my to-do list

Lash-L committed:
Merge branch 'dev' into roborock_hacs_conversion

allenporter committed:
Merge branch 'dev' into roborock_hacs_conversion

allenporter commented:
This says "Merging can be performed automatically once the requested changes are addressed." so it may need re-review by the original requestor? Can't tell.

Lash-L commented:
> This says "Merging can be performed automatically once the requested changes are addressed." so it may need re-review by the original requestor? Can't tell.

Ah yeah - probably because frenck requested changes to begin with but doesn't have a corresponding accepting of said changes. 

I know I'm not supposed to ping, but should we just wait? Will he end up seeing this PR?

allenporter commented:
@frenck the changes you requested (splitting into smaller focused pr were addressed). I've approved otherwise, but the checks would ike your re-review.

Lash-L commented:
Let me know if you need anything from me - I'd love to get it included in the 2023.05 release/ potentially even get a few other small additional features included in the initial release through small PRS, but I completely understand if either one of those things is not possible.

allenporter commented:
I chatted w/ the core members and learned how to dismiss the blocking requested changes since you addressed them a long time ago and it was a trivial requested change. We should be good here now, and can address anything else in follow up PRs.

Lash-L commented:
Amazing! Thanks a ton allen!
